### PR TITLE
validation: take parentRef.port into account

### PR DESCRIPTION
**Pull request by howardjohn**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Currently, we cannot have:

```yaml
  parentRefs:
  - name: gw1
    port: 80
  - name: gw1
    port: 8080
```

This seems like an oversight; our validation only takes sectionName into consideration.

Note: this DOES allow overlapping parentSection selection. For exampe, i may select port=80 and sectionName=http which refer to the same thing. This is not something we can catch in validation. Perhaps we need to clarify the behavior here in the spec, if its not clear. Note this could always happen with 2 different routes anyways, though.

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Review by robscott — COMMENTED** [review]
Thanks @howardjohn! Mostly LGTM, but would be helpful to have one more test case.

File: apis/v1beta1/validation/common_test.go [hunk 1]
```diff
@@ -187,7 +188,67 @@ func TestValidateParentRefs(t *testing.T) {
...
+			},
+			{
+				Name:      "example",
+				Namespace: &namespace,
+				Port:      ptrTo(gatewayv1b1.PortNumber(81)),
+			},
+		},
+		err: "",
+	}, {
+		name: "valid ParentRefs with multiple mixed references to the same parent",
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply apis/v1beta1/validation/common_test.go:208]: Good call. added and the code handled this already

File: apis/v1beta1/validation/common.go [hunk 2]
```diff
@@ -35,30 +35,48 @@ func ValidateParentRefs(parentRefs []gatewayv1b1.ParentReference, path *field.Pa
...
+			pq.section = *p.SectionName
+		}
+		if p.Port != nil {
+			pq.port = *p.Port
 		}
 		if s, ok := parentRefsSectionMap[targetParentRefs]; ok {
-			if len(s[0]) == 0 || len(*targetSection) == 0 {
-				errs = append(errs, field.Required(path.Child("parentRefs"), "sectionNames must be specified when more than one parentRef refers to the same parent"))
+			if s.UnsortedList()[0] == (parentQualifier{}) || pq == (parentQualifier{}) {
+				errs = append(errs, field.Required(path.Child("parentRefs"), "sectionNames or port must be specified when more than one parentRef refers to the same parent"))
```
Comment (gyohuangxin) [inline apis/v1beta1/validation/common.go:60]:
```suggestion
				errs = append(errs, field.Required(path.Child("parentRefs"), "sectionNames or ports must be specified when more than one parentRef refers to the same parent"))
```

**Comment by gyohuangxin** [comment]
LGTM,  but a small typo. @howardjohn Thanks!

**Review by robscott — COMMENTED** [review]
Thanks @howardjohn!

/lgtm
/approve

**Pull request merged**